`coreos.ignition.ssh.key` failing in `kola-azure`

`coreos.ignition.ssh.key` has been failing in the `kola-azure` jenkins job since August 15, 2023 with the following error:
```
[2023-08-17T15:08:35.184Z] --- FAIL: coreos.ignition.ssh.key (331.60s)
[2023-08-17T15:08:35.184Z]         harness.go:1789: mach.Start() failed: machine "kola-975f064f-cccec36d80" failed basic checks: detected failed or stuck systemd units
```
Looking into the journal log file for the test shows:
[journal.txt](https://github.com/coreos/fedora-coreos-tracker/files/12411622/journal.txt)
[console.txt](https://github.com/coreos/fedora-coreos-tracker/files/12411627/console.txt)


```
Aug 17 19:46:17.528507 afterburn[1523]: Aug 17 19:46:17.528 INFO Fetching http://168.63.129.16/?comp=versions: Attempt #1
Aug 17 19:46:17.552714 afterburn[1523]: Aug 17 19:46:17.552 INFO Fetch successful
Aug 17 19:46:17.569624 afterburn[1523]: Aug 17 19:46:17.569 INFO Fetching http://168.63.129.16/machine/?comp=goalstate: Attempt #1
Aug 17 19:46:17.574856 afterburn[1523]: Aug 17 19:46:17.574 INFO Fetch successful
Aug 17 19:46:17.883244 afterburn[1523]: Aug 17 19:46:17.883 INFO Fetching http://168.63.129.16/machine/5c91c37d-f891-4428-93c9-81e602a91a74/a1fc44fd%2Dc611%2D4ea8%2Dbc6b%2D678f2ba3b71e.%5Fkola%2Dc954503e%2D36b448701c?comp=certificates&incarnation=1: Attempt #1
Aug 17 19:46:17.898618 afterburn[1523]: Aug 17 19:46:17.898 INFO Fetch successful
Aug 17 19:46:17.927929 afterburn[1523]: Error: failed to run
Aug 17 19:46:17.927929 afterburn[1523]: Caused by:
Aug 17 19:46:17.927929 afterburn[1523]:     0: writing ssh keys
Aug 17 19:46:17.927929 afterburn[1523]:     1: failed to convert pkcs12 blob to ssh pubkey
Aug 17 19:46:17.927929 afterburn[1523]:     2: failed to get chain from pkcs12
Aug 17 19:46:17.929015 systemd[1]: afterburn-sshkeys@core.service: Main process exited, code=exited, status=1/FAILURE
Aug 17 19:46:17.929193 systemd[1]: afterburn-sshkeys@core.service: Failed with result 'exit-code'.
Aug 17 19:46:17.934176 systemd[1]: Failed to start afterburn-sshkeys@core.service - Afterburn (SSH Keys).
Aug 17 19:46:17.942790 systemd[1]: Dependency failed for afterburn-sshkeys.target - Synchronize afterburn-sshkeys@.service template instances.
Aug 17 19:46:17.947667 systemd[1]: Dependency failed for multi-user.target - Multi-User System.
Aug 17 19:46:17.951000 audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=afterburn-sshkeys@core comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
Aug 17 19:46:17.952533 systemd[1]: multi-user.target: Job multi-user.target/start failed with result 'dependency'.
Aug 17 19:46:17.952960 kernel: audit: type=1130 audit(1692301577.951:80): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=afterburn-sshkeys@core comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
Aug 17 19:46:17.952564 systemd[1]: afterburn-sshkeys.target: Job afterburn-sshkeys.target/start failed with result 'dependency'.
```
Afterburn fails to [convert pkcs12 blob to ssh pubkey](https://github.com/coreos/afterburn/blob/06a80b6edbe2bde448073b9b7f1b171a2507cd65/src/providers/microsoft/azure/mod.rs#L275)
Before this failure, the fetching section looked like this:
```
Aug 14 00:49:48.769452 afterburn[1511]: Aug 14 00:49:48.769 INFO Fetching http://168.63.129.16/?comp=versions: Attempt #1    
Aug 14 00:49:48.790498 afterburn[1511]: Aug 14 00:49:48.790 INFO Fetch successful                                            
Aug 14 00:49:48.809532 afterburn[1511]: Aug 14 00:49:48.809 INFO Fetching http://168.63.129.16/machine/?comp=goalstate: Attem
Aug 14 00:49:48.819160 afterburn[1511]: Aug 14 00:49:48.818 INFO Fetch successful                                            
Aug 14 00:49:48.819462 afterburn[1511]: Aug 14 00:49:48.819 WARN SSH pubkeys requested, but not provisioned for this instance
```
Which is what other test's logs look like when an SSH key IS set. So it's looking like an SSH key is being set somehow, even though one isnt specified. This is causing the `.ssh/authorized_keys.d/afterburn` file to be created and the kola test fails.

The certificates endpoint is now included in `http://168.63.129.16/machine/?comp=goalstate` so now it's being fetched and afterburn is failing directly after [fetching the certificate](https://github.com/coreos/afterburn/blob/06a80b6edbe2bde448073b9b7f1b171a2507cd65/src/providers/microsoft/azure/mod.rs#L266).

--- a/docs/release-notes.md
+++ b/docs/release-notes.md
@@ -13,6 +13,7 @@ Major changes:
 Minor changes:
 
 - Hetzner: fix duplicate attribute prefix
+- Fix Azure SSH key fetching when no key provisioned
 
 Packaging changes:
 

--- a/src/providers/microsoft/azure/mod.rs
+++ b/src/providers/microsoft/azure/mod.rs
@@ -251,7 +251,7 @@ impl Azure {
     }
 
     // put it all together
-    fn get_ssh_pubkey(&self, certs_endpoint: String) -> Result<PublicKey> {
+    fn get_ssh_pubkey(&self, certs_endpoint: String) -> Result<Option<PublicKey>> {
         // we have to generate the rsa public/private keypair and the x509 cert
         // that we use to make the request. this is equivalent to
         // `openssl req -x509 -nodes -subj /CN=LinuxTransport -days 365 -newkey rsa:2048 -keyout private.pem -out cert.pem`
@@ -271,10 +271,7 @@ impl Azure {
             .context("failed to decrypt cms blob")?;
 
         // convert that to the OpenSSH public key format
-        let ssh_pubkey = crypto::p12_to_ssh_pubkey(&p12)
-            .context("failed to convert pkcs12 blob to ssh pubkey")?;
-
-        Ok(ssh_pubkey)
+        crypto::p12_to_ssh_pubkey(&p12).context("failed to convert pkcs12 blob to ssh pubkey")
     }
 
     #[cfg(test)]
@@ -402,18 +399,17 @@ impl MetadataProvider for Azure {
         let goalstate = self.fetch_goalstate()?;
         let certs_endpoint = match goalstate.certs_endpoint() {
             Some(ep) => ep,
-            None => {
-                warn!("SSH pubkeys requested, but not provisioned for this instance");
-                return Ok(vec![]);
-            }
+            None => return Ok(vec![]),
         };
 
         if certs_endpoint.is_empty() {
             bail!("unexpected empty certificates endpoint");
         }
 
-        let key = self.get_ssh_pubkey(certs_endpoint)?;
-        Ok(vec![key])
+        let maybe_key = self.get_ssh_pubkey(certs_endpoint)?;
+        let key: Vec<PublicKey> = maybe_key.into_iter().collect();
+
+        Ok(key)
     }
 
     fn boot_checkin(&self) -> Result<()> {

--- a/src/providers/microsoft/azurestack/mod.rs
+++ b/src/providers/microsoft/azurestack/mod.rs
@@ -267,7 +267,7 @@ impl AzureStack {
     }
 
     // put it all together
-    fn get_ssh_pubkey(&self, certs_endpoint: String) -> Result<PublicKey> {
+    fn get_ssh_pubkey(&self, certs_endpoint: String) -> Result<Option<PublicKey>> {
         // we have to generate the rsa public/private keypair and the x509 cert
         // that we use to make the request. this is equivalent to
         // `openssl req -x509 -nodes -subj /CN=LinuxTransport -days 365 -newkey rsa:2048 -keyout private.pem -out cert.pem`
@@ -287,10 +287,7 @@ impl AzureStack {
             .context("failed to decrypt cms blob")?;
 
         // convert that to the OpenSSH public key format
-        let ssh_pubkey = crypto::p12_to_ssh_pubkey(&p12)
-            .context("failed to convert pkcs12 blob to ssh pubkey")?;
-
-        Ok(ssh_pubkey)
+        crypto::p12_to_ssh_pubkey(&p12).context("failed to convert pkcs12 blob to ssh pubkey")
     }
 
     fn fetch_hostname(&self) -> Result<Option<String>> {
@@ -326,18 +323,17 @@ impl MetadataProvider for AzureStack {
         let goalstate = self.fetch_goalstate()?;
         let certs_endpoint = match goalstate.certs_endpoint() {
             Some(ep) => ep,
-            None => {
-                warn!("SSH pubkeys requested, but not provisioned for this instance");
-                return Ok(vec![]);
-            }
+            None => return Ok(vec![]),
         };
 
         if certs_endpoint.is_empty() {
             bail!("unexpected empty certificates endpoint");
         }
 
-        let key = self.get_ssh_pubkey(certs_endpoint)?;
-        Ok(vec![key])
+        let maybe_key = self.get_ssh_pubkey(certs_endpoint)?;
+        let key: Vec<PublicKey> = maybe_key.into_iter().collect();
+
+        Ok(key)
     }
 
     fn boot_checkin(&self) -> Result<()> {

--- a/src/providers/microsoft/crypto/mod.rs
+++ b/src/providers/microsoft/crypto/mod.rs
@@ -54,7 +54,7 @@ pub fn decrypt_cms(smime: &[u8], pkey: &PKey<Private>, x509: &X509) -> Result<Ve
     Ok(p12_der)
 }
 
-pub fn p12_to_ssh_pubkey(p12_der: &[u8]) -> Result<PublicKey> {
+pub fn p12_to_ssh_pubkey(p12_der: &[u8]) -> Result<Option<PublicKey>> {
     // the contents of that encrypted cms blob we got are actually a different
     // cryptographic structure. we read that in from the contents and parse it.
     // PKCS12 has the ability to have a password, but we don't have one, hence
@@ -65,7 +65,14 @@ pub fn p12_to_ssh_pubkey(p12_der: &[u8]) -> Result<PublicKey> {
     // ParsedPKCS12_2 has three parts: a pkey, a main x509 cert, and a list of other
     // x509 certs. The list of other x509 certs are called the `certificate chain`
     // currently denoted as `ca`; there is only one cert in this `certificate chain`,
-    // which is the ssh public key.
+    // which is the ssh public key. The certs endpoint may still be populated even if
+    // an SSH public key hasn't been provided, which would lead to this code being
+    // executed. The certificate chain will be empty in this case, so just return
+    // None and handle it further up the stack.
+    if p12.ca.is_none() {
+        return Ok(None);
+    }
+
     let ca = p12
         .ca
         .ok_or_else(|| anyhow!("failed to get chain from pkcs12"))?;
@@ -86,5 +93,5 @@ pub fn p12_to_ssh_pubkey(p12_der: &[u8]) -> Result<PublicKey> {
     let n = ssh_pubkey_rsa.n().to_vec();
     let ssh_pubkey = PublicKey::from_rsa(e, n);
 
-    Ok(ssh_pubkey)
+    Ok(Some(ssh_pubkey))
 }

